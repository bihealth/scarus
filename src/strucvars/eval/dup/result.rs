//! Data structures for representing the actual results of DEL/copy number gain.

use crate::strucvars::{
    data::{clingen_dosage, hgnc::GeneIdInfo},
    eval::{
        common::{FunctionalElement, GeneOverlap, ScoreRange, SuggestedScore},
        result::{ClinvarSvOverlap, GnomadSvOverlap, Pvs1Result},
    },
};

/// Evaluation results for each section of the ACMG rule.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum Section {
    /// Results of Section G1.
    G1(G1),
    /// Results of Section G2.
    G2(G2),
    /// Results of Section G3.
    G3(G3),
    /// Results of Section G4.
    G4(G4),
}

impl SuggestedScore for Section {
    fn suggested_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(g1a)) => g1a.suggested_score(),
            Section::G1(G1::G1B(g1b)) => g1b.suggested_score(),
            Section::G2(G2::G2A(g2a)) => g2a.suggested_score(),
            Section::G2(G2::G2B(g2b)) => g2b.suggested_score(),
            Section::G2(G2::G2C(g2c)) => g2c.suggested_score(),
            Section::G2(G2::G2D(g2d)) => g2d.suggested_score(),
            Section::G2(G2::G2E(g2e)) => g2e.suggested_score(),
            Section::G2(G2::G2F(g2f)) => g2f.suggested_score(),
            Section::G2(G2::G2G(g2g)) => g2g.suggested_score(),
            Section::G2(G2::G2H(g2h)) => g2h.suggested_score(),
            Section::G2(G2::G2I(g2i)) => g2i.suggested_score(),
            Section::G2(G2::G2J(g2j)) => g2j.suggested_score(),
            Section::G2(G2::G2L(g2l)) => g2l.suggested_score(),
            Section::G3(G3::G3A(g3a)) => g3a.suggested_score(),
            Section::G3(G3::G3B(g3b)) => g3b.suggested_score(),
            Section::G3(G3::G3C(g3c)) => g3c.suggested_score(),
            Section::G4(G4::G4Patho(g4p)) => g4p.suggested_score(),
            Section::G4(G4::G4N(g4n)) => g4n.suggested_score(),
            Section::G4(G4::G4O(g4o)) => g4o.suggested_score(),
        }
    }
}

impl ScoreRange for Section {
    fn min_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(_)) => 0.0,
            Section::G1(G1::G1B(_)) => -0.6,
            Section::G2(G2::G2A(_)) => 1.0,
            Section::G2(G2::G2B(_)) => 0.0,
            Section::G2(G2::G2C(_)) => -1.0,
            Section::G2(G2::G2D(_)) => -1.0,
            Section::G2(G2::G2E(_)) => 0.0,
            Section::G2(G2::G2F(_)) => -1.0,
            Section::G2(G2::G2G(_)) => 0.0,
            Section::G2(G2::G2H(_)) => 0.0,
            Section::G2(G2::G2I(_)) => 0.0,
            Section::G2(G2::G2J(_)) => 0.0,
            Section::G2(G2::G2L(_)) => 0.0,
            Section::G3(G3::G3A(_)) => 0.0,
            Section::G3(G3::G3B(_)) => 0.45,
            Section::G3(G3::G3C(_)) => 0.9,
            Section::G4(G4::G4Patho(_)) => -0.9,
            Section::G4(G4::G4N(_)) => -0.9,
            Section::G4(G4::G4O(_)) => -1.0,
        }
    }

    fn max_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(_)) => 0.0,
            Section::G1(G1::G1B(_)) => -0.6,
            Section::G2(G2::G2A(_)) => 1.0,
            Section::G2(G2::G2B(_)) => 0.0,
            Section::G2(G2::G2C(_)) => -1.0,
            Section::G2(G2::G2D(_)) => -1.0,
            Section::G2(G2::G2E(_)) => 0.0,
            Section::G2(G2::G2F(_)) => 0.0,
            Section::G2(G2::G2G(_)) => 0.0,
            Section::G2(G2::G2H(_)) => 0.0,
            Section::G2(G2::G2I(_)) => 0.9,
            Section::G2(G2::G2J(_)) => 0.0,
            Section::G2(G2::G2L(_)) => 0.0,
            Section::G3(G3::G3A(_)) => 0.0,
            Section::G3(G3::G3B(_)) => 0.45,
            Section::G3(G3::G3C(_)) => 0.9,
            Section::G4(G4::G4Patho(_)) => 0.9,
            Section::G4(G4::G4N(_)) => 0.9,
            Section::G4(G4::G4O(_)) => 0.0,
        }
    }
}

/// Enumeration of the categories for the copy number gain evaluation, Section 1.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G1 {
    /// Contains protein-coding or other known functionally important elements.
    G1A(G1A),
    /// Does NOT contain protein-coding or any functionally important elements.
    G1B(G1B),
}

/// Result of the G1A subsection (important feature).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G1A {
    /// Overlapping transcripts/genes.
    pub genes: Vec<GeneOverlap>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G1A {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Result of the G1B subsection (no important feature).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G1B {
    // no members
}

impl SuggestedScore for G1B {
    fn suggested_score(&self) -> f32 {
        -0.6
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 2.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G2 {
    /// Complete overlap; Tthe TS gene or minimal criticala reegion is fully contained
    /// within the observed copy number gain.
    G2A(G2A),
    /// Partial overlap of an established TS region.
    G2B(G2B),
    /// Identical in gene content to the established benign copy number gain.
    G2C(G2C),
    /// Smaller than established benign copy number gain, breakpoint(s) does not
    /// interrupt protein-coding genes.
    G2D(G2D),
    /// Smaller than established benign copy number gain, breakpoint(s) potentially
    /// interrupt protein-coding genes.
    G2E(G2E),
    /// Larger tahn known benign copy-number gain, does not include additional
    /// protein-coding genes.
    G2F(G2F),
    /// Overlaps a benign copy number gain but includes additional material.
    G2G(G2G),
    /// TS gene fully contained within obsrved copy-number gain.
    G2H(G2H),
    /// Both breakpoints are within the asme gene (gene-level sequence variant, possibly
    /// resulting in LoF.
    G2I(G2I),
    /// One breakpoint is within an established TS gene, patient's phenotype is either
    /// inconsistent with what is expecte for LOF or that gene OR unknown
    G2J(G2J),
    /// One or both breakpoints are within gene(s) of no established clinical significance.
    G2L(G2L),
}

/// Result of the G2A subsection (complete overlap).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2A {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping TS genes.
    pub ts_genes: Vec<clingen_dosage::Gene>,
    /// Overlapping TS genomic regions.
    pub ts_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2A {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2B subsection (partial overlap of established TS region).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2B {
    /// Overlapping TS genomic regions.
    pub ts_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2B {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Result of the G2C subsection (identical in gene content to the established
/// benign TS region).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2C {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping TS genomic regions.
    pub benign_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2C {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2D subsection (smaller than established benign copy-number gain,
/// breakpoint(s) does not interrupt protein-coding gene).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2D {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic region.
    pub benign_region: clingen_dosage::Region,
}

impl SuggestedScore for G2D {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2E subsection (smaller than established benign copy-number gain,
/// breakpoint(s) potentially interrupt protein-coding gene).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2E {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
    /// Potentially affected genes.
    pub genes: Vec<GeneIdInfo>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G2E {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of G2F subsection (larger than known benign copy-number gain,
/// does not include additional protein-coding genes).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2F {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
}

impl SuggestedScore for G2F {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2G subsection (overlaps a benign copy-number gain but includes
/// additional genomic material).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2G {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
    /// Additional genes.
    pub genes: Vec<GeneIdInfo>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G2G {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2H subsection (HI gene fully contained within observed copy-number
/// gain).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2H {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping HI gene information.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2H {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2I subsection (both breakpoint are within the same gene, potential
/// LOF, fall back to PVS1).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2I {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// PVS assessment result.
    pub pvs1_result: Pvs1Result,
    /// HI genes for which the assessment was made.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2I {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2J subsection (One breakpoint is within an esbalished HI gene,
/// patient's phenotype is either inconsistent with what is expected for LOF of
/// that gene OR unknown).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2J {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// HI genes overlapping with the CNV.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2J {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2L subsection (one or both breakpoints are within gene(s) of
/// no established linical significance).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2L {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Genes overlapping with the DUP.
    pub genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2L {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 2.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G3 {
    /// <=34 genes.
    G3A(G3Count),
    /// 35-49 genes.
    G3B(G3Count),
    /// >=50 genes.
    G3C(G3Count),
}

/// Result of the 3A subsection (Number of protein-coding RefSeq genes wholly or partially included
/// in the copy-number loss).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G3Count {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Number of protein-coding RefSeq genes wholly or partially included in the copy-number loss.
    pub num_genes: usize,
    /// The overlapping genes.
    pub genes: Vec<GeneOverlap>,
}

impl SuggestedScore for G3Count {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 4.
///
/// Only 4O can be automatically determined.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G4 {
    /// Overlap with pathogenic variants; must be evaluated by a human.
    ///
    /// This could be one of 4A, 4B, 4C, 4D, 4E, 4L, 4M.
    G4Patho(G4Patho),
    /// Overlap with benign variants; must be evaluated by a human.
    ///
    /// This is roughly equiavalent to 4N.
    G4N(G4N),
    /// Case–control and population evidence; Overlap with common population variation.
    G4O(G4O),
}

/// Result of the 4O subsection (overlap with common population variation).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4Patho {
    /// Overlapping ClinVar variants, descendingly by overlap.
    pub overlaps: Vec<ClinvarSvOverlap>,
}

impl SuggestedScore for G4Patho {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Benign variants from ClinVar.
///
/// This is roughly 4N.
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4N {
    /// Overlapping ClinVar variants, descendingly by overlap.
    pub overlaps: Vec<ClinvarSvOverlap>,
}

impl SuggestedScore for G4N {
    fn suggested_score(&self) -> f32 {
        -0.9
    }
}

/// Result of the 4O subsection (overlap with common population variation).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4O {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Accession identifiers of overlapping common variants.
    pub overlaps: Vec<GnomadSvOverlap>,
}

impl SuggestedScore for G4O {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}
