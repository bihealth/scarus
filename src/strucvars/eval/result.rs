//! Common code for evaluation results of CNVs.

use annonars::gnomad_sv::cli::query::Record as GnomadSvRecord;
use annonars::pbs::clinvar::sv::Record as ClinvarSvRecord;
use intervals_general::bound_pair::BoundPair;

use super::{del, dup};

/// Overall result.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum Evaluation {
    /// Result of assessing DUPs.
    Dup(dup::result::Evaluation),
    /// Result of assessing DELs.
    Del(del::result::Evaluation),
}

/// Overal assessment of a CNV.
#[derive(
    Debug,
    Clone,
    Copy,
    Default,
    PartialEq,
    strum_macros::EnumIter,
    serde::Deserialize,
    serde::Serialize,
)]
#[serde(rename_all = "snake_case")]
pub enum ClinicalSignificance {
    /// Benign
    Benign,
    /// Likely benign
    LikelyBenign,
    /// Uncertain_significance
    #[default]
    UncertainSignificance,
    /// Likely_pathogenic
    LikelyPathogenic,
    /// Pathogenic
    Pathogenic,
}

/// Score interval to use.
pub type ScoreRange = intervals_general::interval::Interval<f32>;

impl ClinicalSignificance {
    /// Return the scores,
    pub fn range(self) -> ScoreRange {
        match self {
            ClinicalSignificance::Benign => ScoreRange::UnboundedClosedRight { right: 0.99 },
            ClinicalSignificance::LikelyBenign => ScoreRange::LeftHalfOpen {
                bound_pair: BoundPair::new(-0.99, -0.90).expect("invalid interval?"),
            },
            ClinicalSignificance::UncertainSignificance => ScoreRange::Open {
                bound_pair: BoundPair::new(-0.90, 0.90).expect("invalid interval?"),
            },
            ClinicalSignificance::LikelyPathogenic => ScoreRange::LeftHalfOpen {
                bound_pair: BoundPair::new(0.90, 0.99).expect("invalid interval?"),
            },
            ClinicalSignificance::Pathogenic => ScoreRange::UnboundedClosedLeft { left: 0.99 },
        }
    }
}

/// Enumeration describing the PVS1 results.
#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Default,
    serde::Deserialize,
    serde::Serialize,
)]
#[serde(rename_all = "snake_case")]
pub enum Pvs1Result {
    /// PVS1
    #[default]
    Pvs1,
    /// PVS1_Strong
    Pvs1Strong,
    /// PVS1_Moderate
    Pvs1Moderate,
    /// PVS1_Supporting
    Pvs1Supporting,
}

/// Information about one overlapping ClinVar records.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct ClinvarSvOverlap {
    /// Reciprocal overlap.
    pub overlap: f32,
    /// Overlapping ClinVar record.
    pub record: ClinvarSvRecord,
}

/// Information about one overlapping gnomAD records.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct GnomadSvOverlap {
    /// Reciprocal overlap.
    pub overlap: f32,
    /// Overlapping ClinVar record.
    pub record: GnomadSvRecord,
}
