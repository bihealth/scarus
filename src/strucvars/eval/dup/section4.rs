//! Implementation of evaluation of copy number loss section 4.
//!
//! Note that we can only determine 4O reliably in an automated fashion.  We support
//! reporting overlapping variants from ClinVar but leave them as "dangle" because a
//! human must evaluate the phenotype.

use super::result::{G4Patho, Section, G4, G4N};
use crate::strucvars::{
    ds::StructuralVariant,
    eval::{
        dup::result::G4O,
        result::{ClinvarSvOverlap, GnomadSvOverlap},
        CarrierFrequency, IntoInterval, Overlaps, SvType,
    },
};
use annonars::{
    clinvar_sv::cli::query::EXAC_CNV_CASES,
    gnomad_sv::cli::query::Record as GnomadSvRecord,
    pbs::clinvar::{
        minimal::{ClinicalSignificance, ReviewStatus, VariantType},
        sv::Record as ClinvarSvRecord,
    },
};
use bio::bio_types::genome::{AbstractInterval, Interval};

/// Evaluation of duplications, gain of copy number.
///
/// This is mainly used to encapsulate the functionality.  Creating new such
/// objects is very straightforward and cheap.
pub struct Evaluator<'a> {
    /// The parent evaluator.
    #[allow(dead_code)] // TODO
    parent: &'a super::super::Evaluator,
}

impl<'a> Evaluator<'a> {
    /// Create a new `Evaluator`.
    pub fn with_parent(parent: &'a super::super::Evaluator) -> Self {
        Self { parent }
    }

    /// Perform the evaluation of copy number loss Section 3 and all subsection.
    ///
    /// # Arguments
    ///
    /// * `strucvar` - Structural variant to be evaluated.
    ///
    /// # Returns
    ///
    /// The evaluation result.
    ///
    /// # Errors
    ///
    /// If anything goes wrong, it returns a generic `anyhow::Error`.
    pub fn evaluate(&self, strucvar: &StructuralVariant) -> Result<Vec<Section>, anyhow::Error> {
        tracing::debug!("evaluating section 4 for {:?}", strucvar);
        let sv_interval: Interval = strucvar.clone().into();
        let mut result = Vec::new();

        // Retrieve all overlapping ClinVar records.
        let clinvar_records = self.overlapping_clinvar_records(&sv_interval)?;

        if let Some(result_clinvar_patho) =
            self.handle_clinvar_pathogenic(&sv_interval, &clinvar_records)?
        {
            result.push(result_clinvar_patho)
        }
        if let Some(result_clinvar_benign) =
            self.handle_clinvar_benign(&sv_interval, &clinvar_records)?
        {
            result.push(result_clinvar_benign)
        }
        if let Some(result_gnomad_sv) = self.handle_gnomad_sv(&sv_interval)? {
            result.push(result_gnomad_sv)
        }

        tracing::warn!("Section 4 evaluation not implemented yet");
        Ok(result)
    }

    /// Obtain overlapping ClinVar records.
    fn overlapping_clinvar_records(
        &self,
        sv_interval: &Interval,
    ) -> Result<Vec<ClinvarSvRecord>, anyhow::Error> {
        tracing::debug!("querying ClinVar for {:?}", sv_interval);
        let res = self
            .parent
            .clinvar_sv_overlaps(
                sv_interval.contig(),
                sv_interval.range().start as u32,
                sv_interval.range().end as u32,
            )
            .map_err(|e| anyhow::anyhow!("ClinVar query failed: {}", e))?;
        tracing::debug!("found {} overlapping ClinVar records", res.len());
        tracing::trace!("records are {:#?}", &res);
        Ok(res)
    }

    /// Evaluate ClinVar pathogenic variants.
    ///
    /// This could be one of 4A, 4B, 4C, 4D, 4E, 4L, 4M.
    fn handle_clinvar_pathogenic(
        &self,
        sv_interval: &Interval,
        clinvar_records: &[ClinvarSvRecord],
    ) -> Result<Option<Section>, anyhow::Error> {
        let pathogenic = [
            ClinicalSignificance::Pathogenic as i32,
            ClinicalSignificance::LikelyPathogenic as i32,
        ];
        // Filter intervals to pathogenic ones without conflicts and sufficient overlap.
        let mut overlaps = clinvar_records
            .iter()
            .flat_map(|record| {
                let rc = record.reference_assertions.first().expect("no RCV");
                let record_interval = record.into_interval();
                let overlap = sv_interval.reciprocal_overlap(&record_interval) as f32;
                tracing::trace!("considering records {:?} with overlap {}", record, overlap);
                tracing::trace!("rc.clinical_significance = {} / {:?}", rc.clinical_significance, &pathogenic);
                let is_pathogenic = pathogenic.contains(&rc.clinical_significance);
                let is_review_status = rc.review_status <= ReviewStatus::CriteriaProvidedSingleSubmitter as i32;
                let is_variant_type = record.variant_type == VariantType::Duplication as i32;
                let is_overlap = overlap >= self.parent.config.clinvar_sv_min_overlap_pathogenic;
                let is_interesting = is_pathogenic && is_variant_type && is_overlap;
                tracing::trace!("is_pathogenic: {}, is_review_status: {}, is_variant_type: {}, is_overlap: {}, is_interesting: {}", is_pathogenic, is_review_status, is_variant_type, is_overlap, is_interesting);
                if is_interesting {
                    Some(ClinvarSvOverlap {
                        overlap,
                        record: record.clone(),
                    })
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();

        overlaps.sort_by(|a, b| {
            b.overlap
                .partial_cmp(&a.overlap)
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        if overlaps.is_empty() {
            Ok(None)
        } else {
            Ok(Some(Section::G4(G4::G4Patho(G4Patho { overlaps }))))
        }
    }

    /// Evaluate ClinVar benign variants.
    ///
    /// This is roughly 4N.
    fn handle_clinvar_benign(
        &self,
        sv_interval: &Interval,
        clinvar_records: &[ClinvarSvRecord],
    ) -> Result<Option<Section>, anyhow::Error> {
        let benign = [
            ClinicalSignificance::Benign as i32,
            ClinicalSignificance::LikelyBenign as i32,
        ];
        // Filter intervals to pathogenic ones without conflicts and sufficient overlap.
        let mut overlaps = clinvar_records
            .iter()
            .flat_map(|record| {
                let rc = record.reference_assertions.first().expect("no RCV");
                let record_interval = record.into_interval();
                let overlap = sv_interval.reciprocal_overlap(&record_interval) as f32;
                tracing::trace!("considering records {:?} with overlap {}", record, overlap);
                let interesting = benign.contains(&rc.clinical_significance)
                    && rc.review_status <= ReviewStatus::CriteriaProvidedSingleSubmitter as i32
                    && record.variant_type == VariantType::Duplication as i32
                    && overlap >= self.parent.config.clinvar_sv_min_overlap_pathogenic;
                if interesting {
                    Some(ClinvarSvOverlap {
                        overlap,
                        record: record.clone(),
                    })
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();

        overlaps.sort_by(|a, b| {
            b.overlap
                .partial_cmp(&a.overlap)
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        if overlaps.is_empty() {
            Ok(None)
        } else {
            Ok(Some(Section::G4(G4::G4N(G4N { overlaps }))))
        }
    }

    /// Handle overlap in gnomAD-SV.
    ///
    /// This corresponds to case 4O.
    fn handle_gnomad_sv(&self, sv_interval: &Interval) -> Result<Option<Section>, anyhow::Error> {
        // Query gnomAD-SV for all records.
        tracing::debug!("querying gnomAD-SV for {:?}", sv_interval);
        let raw_records = self
            .parent
            .gnomad_sv_overlaps(
                sv_interval.contig(),
                sv_interval.range().start as u32,
                sv_interval.range().end as u32,
            )
            .map_err(|e| anyhow::anyhow!("gnomAD-SV query failed: {}", e))?;
        tracing::debug!("found {} overlapping gnomAD-SV records", raw_records.len());
        tracing::trace!("records are {:#?}", &raw_records);

        // Filter to records with the correct type and appropriate overlap.
        let mut overlaps: Vec<GnomadSvOverlap> = raw_records
            .into_iter()
            .flat_map(|record| {
                let record_interval = record.into_interval();
                let overlap = sv_interval.reciprocal_overlap(&record_interval) as f32;
                let sv_type = SvType::try_from(&record);
                let interesting = matches!(sv_type, Ok(SvType::Dup | SvType::Cnv))
                    && overlap >= self.parent.config.gnomad_sv_min_overlap;
                if interesting {
                    Some(GnomadSvOverlap { overlap, record })
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();
        overlaps.sort_by(|a, b| {
            b.overlap
                .partial_cmp(&a.overlap)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        tracing::trace!("filtered overlaps are {:#?}", &overlaps);

        // We have to manually count ExAC CNV records as they are per-sample and not aggregated.
        let exac_cnv_count = overlaps
            .iter()
            .filter(|GnomadSvOverlap { record, .. }| matches!(record, &GnomadSvRecord::ExacCnv(_)))
            .count();
        let exac_freq = exac_cnv_count as f32 / EXAC_CNV_CASES as f32;
        // For the rest, we can obtain the frequency from the records.
        let other_freqs = overlaps
            .iter()
            .flat_map(|GnomadSvOverlap { record, .. }| {
                if matches!(record, &GnomadSvRecord::ExacCnv(_)) {
                    None
                } else {
                    Some(record.carrier_freq().expect("no carrier freq?"))
                }
            })
            .collect::<Vec<_>>();
        let max_freq = other_freqs
            .iter()
            .max_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal))
            .copied()
            .unwrap_or(0.0);

        let suggested_score = if exac_freq >= self.parent.config.gnomad_sv_min_freq_common
            || max_freq >= self.parent.config.gnomad_sv_min_freq_common
        {
            -1.0
        } else {
            0.0
        };

        Ok(if overlaps.is_empty() {
            None
        } else {
            Some(Section::G4(G4::G4O(G4O {
                suggested_score,
                overlaps,
            })))
        })
    }
}

#[cfg(test)]
mod test {
    use crate::strucvars::ds::{StructuralVariant, SvType};

    use super::super::super::test::global_evaluator_37;

    #[tracing_test::traced_test]
    #[rstest::rstest]
    // ClinVar pathogenic variant
    #[case("10", 55_581_618, 56_288_162, "VCV000806586")]
    // ClinVar benign variant
    #[case("10", 135_252_347, 135_378_802, "VCV000221841")]
    // gnomAD SV variants
    // ExAC CNV
    #[case("10", 92_994, 95_183, "chr10-92994-95183-NFE")]
    // gnomAD-SV v2
    #[case("1", 157_000, 166_000, "gnomAD-SV_v2.1_DUP_1_6")]
    fn test_evaluate(
        #[case] chrom: &str,
        #[case] start: u32,
        #[case] stop: u32,
        #[case] label: &str,
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", label);

        let evaluator = super::Evaluator::with_parent(global_evaluator_37);

        let strucvar = StructuralVariant {
            chrom: chrom.into(),
            start,
            stop,
            svtype: SvType::Dup,
            ambiguous_range: None,
        };

        insta::assert_yaml_snapshot!(evaluator.evaluate(&strucvar)?);

        Ok(())
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    // same as in `evaluate` test above, keep in sync
    #[case("10", 55_581_618, 56_288_162, "VCV000806586")]
    fn handle_clinvar_pathogenic(
        #[case] chrom: &str,
        #[case] start: u32,
        #[case] stop: u32,
        #[case] label: &str,
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", label);

        let evaluator = super::Evaluator::with_parent(global_evaluator_37);
        let strucvar = StructuralVariant {
            chrom: chrom.into(),
            start,
            stop,
            svtype: SvType::Dup,
            ambiguous_range: None,
        };
        let sv_interval = strucvar.into();

        let clinvar_records = evaluator.overlapping_clinvar_records(&sv_interval)?;

        insta::assert_yaml_snapshot!(
            evaluator.handle_clinvar_pathogenic(&sv_interval, &clinvar_records)?
        );

        Ok(())
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    // same as in `evaluate` test above, keep in sync
    #[case("10", 135_252_347, 135_378_802, "VCV000221841")]
    fn handle_clinvar_benign(
        #[case] chrom: &str,
        #[case] start: u32,
        #[case] stop: u32,
        #[case] label: &str,
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", label);

        let evaluator = super::Evaluator::with_parent(global_evaluator_37);
        let strucvar = StructuralVariant {
            chrom: chrom.into(),
            start,
            stop,
            svtype: SvType::Dup,
            ambiguous_range: None,
        };
        let sv_interval = strucvar.into();

        let clinvar_records = evaluator.overlapping_clinvar_records(&sv_interval)?;

        insta::assert_yaml_snapshot!(
            evaluator.handle_clinvar_benign(&sv_interval, &clinvar_records)?
        );

        Ok(())
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    // same as in `evaluate` test above, keep in sync
    // ExAC CNV
    #[case("10", 92_994, 95_183, "chr10-92994-95183-NFE")]
    // gnomAD-SV v2
    #[case("1", 157_000, 166_000, "gnomAD-SV_v2.1_DUP_1_6")]
    fn handle_gnomad_sv(
        #[case] chrom: &str,
        #[case] start: u32,
        #[case] stop: u32,
        #[case] label: &str,
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", label);

        let evaluator = super::Evaluator::with_parent(global_evaluator_37);
        let strucvar = StructuralVariant {
            chrom: chrom.into(),
            start,
            stop,
            svtype: SvType::Dup,
            ambiguous_range: None,
        };
        let sv_interval = strucvar.into();

        insta::assert_yaml_snapshot!(evaluator.handle_gnomad_sv(&sv_interval)?);

        Ok(())
    }
}
