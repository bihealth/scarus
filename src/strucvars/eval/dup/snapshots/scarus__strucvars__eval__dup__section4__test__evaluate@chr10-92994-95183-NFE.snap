---
source: src/strucvars/eval/dup/section4.rs
expression: evaluator.evaluate(&strucvar)?
---
- G4:
    G4O:
      suggested_score: 0
      overlaps:
        - overlap: 1
          record:
            exac_cnv:
              chrom: "10"
              start: 92994
              stop: 95183
              sv_type: 1
              population: 3
        - overlap: 1
          record:
            exac_cnv:
              chrom: "10"
              start: 92994
              stop: 95183
              sv_type: 1
              population: 3

